File: extractor.go
package main

import (
	"encoding/json"
)

type StateMachineDefinition struct {
	Identifier string `json:"identifier"`
	Definition string `json:"definition"`
}

func extractDefinition(content string) ([]StateMachineDefinition, error) {
	jsonBody := make(map[string]interface{})
	err := json.Unmarshal([]byte(content), &jsonBody)

	if err != nil {
		return nil, err
	}

	for k, v := range jsonBody {
		if k == "Resources" {
			body := v.(map[string]interface{})

			if body != nil {
				filtered := filterStateMachines(body)
				definitions := resolveStateMachineDefinitions(filtered)
				return definitions, nil
			}
		}
	}

	return nil, nil
}

func filterStateMachines(r map[string]interface{}) map[string]map[string]interface{} {
	filtered := make(map[string]map[string]interface{})

	for key, v := range r {
		resource := v.(map[string]interface{})
		if resource == nil {
			continue
		}

		if val, ok := resource["Type"]; ok {
			typeString := val.(string)
			if typeString == "AWS::StepFunctions::StateMachine" {
				if resourceVal, resourceOk := resource["Properties"]; resourceOk {
					props := resourceVal.(map[string]interface{})
					if props != nil {
						filtered[key] = props
					}
				}
			}
		}
	}

	return filtered
}

func resolveStateMachineDefinitions(properties map[string]map[string]interface{}) []StateMachineDefinition {
	var definitions []StateMachineDefinition

	for key, value := range properties {
		if val, ok := value["DefinitionString"]; ok {
			definitions = append(definitions, StateMachineDefinition{
				Identifier: key,
				Definition: val.(string),
			})
		}
	}

	return definitions
}

File: src/extractor.ts
import { StateMachineDefinition } from "./template";

/**
 * Pulls the definitions out of the CloudFormation JSON Template file that is supplied
 * @param contents the String representation of the JSON Template File
 * @returns an array of StateMachineDefintion Objects
 */
export const extractDefinition = (
    contents: string
): StateMachineDefinition[] => {
    let smd: StateMachineDefinition[] = [];

    const parsed = JSON.parse(contents);

    if (!parsed.hasOwnProperty("Resources")) {
        throw new Error("Resources key does not exist on template");
    }

    return filterStateMachines(parsed.Resources);
};

/**
 * Filters out the non-needed elements of the template.  Specifically looking for the
 * `AWS::StepFunctions::StateMachine` Objects
 * @param resources Representation of the "Resources" object inside the JSON Template
 * @returns the list of StateMachineDefinition objects that are found
 */
export const filterStateMachines = (
    resources: any
): StateMachineDefinition[] => {
    let stateMachines: StateMachineDefinition[] = [];
    Object.keys(resources).forEach((element: any) => {
        const r = resources[element];
        if (r.hasOwnProperty("Type")) {
            if (r.Type === "AWS::StepFunctions::StateMachine") {
                stateMachines.push({
                    identifier: element as string,
                    definition: extractAsl(r),
                });
            }
        }
    });

    return stateMachines;
};

/**
 * Extracts the AWS State Language or ASL from the DefinitionString of the Resource
 * @param resource  the AWS Resource to extract the DefintionString from
 * @returns the ASL as defined in the DefinitionString
 */
export const extractAsl = (resource: any): string => {
    if (!resource.hasOwnProperty("Properties")) {
        throw new Error("Resource structure is missing Properties");
    }

    if (!resource.Properties.hasOwnProperty("DefinitionString")) {
        throw new Error("Properties structure is missing DefinitionString");
    }

    return resource.Properties.DefinitionString as string;
};
